error/s2n_errno.h

```diff
@@ -218,5 +219,28 @@ extern __thread const char *s2n_debug_str;
 #define S2N_ERROR_IF( cond , x ) do { if ( cond ) { S2N_ERROR( x ); }} while (0)
 #define S2N_ERROR_IF_PTR( cond , x ) do { if ( cond ) { S2N_ERROR_PTR( x ); }} while (0)
 
+#define S2N_PRECONDITION( cond ) S2N_ERROR_IF(!(cond), S2N_ERR_PRECONDITION_VIOLATION)
+#define S2N_PRECONDITION_PTR( cond ) S2N_ERROR_IF_PTR(!(cond), S2N_ERR_PRECONDITION_VIOLATION)
+
+/**
+ * Define function contracts.
+ * When the code is being verified using CBMC these contracts are formally verified;
+ * When the code is built in debug mode, they are checked as much as possible using assertions
+ * When the code is built in production mode, non-fatal contracts are not checked.
+ * Violations of the function contracts are undefined behaviour.
+ */
+#ifdef CBMC
+#    define S2N_MEM_IS_READABLE(base, len) __CPROVER_r_ok((base), (len))
+#    define S2N_MEM_IS_WRITABLE(base, len) __CPROVER_w_ok((base), (len))
+#else
+/* the C runtime does not give a way to check these properties,
+ * but we can at least check that the pointer is valid */
+#    define S2N_MEM_IS_READABLE(base, len) (((len) == 0) || (base))
```

Comment:
I haven't actually built s2n, but have you checked that these macros are preserved when compiling in debug mode? This is a side-effect free expression. I just wrote some small tests, and Clang is aggressive about compiling out such expressions, even when you compile with `-O0`.

I think this isn't an issue if you always use the macro within a conditional (`if(S2N_MEM_IS_READABLE(foo, 5){ ... }`), but if you intend to e.g. put this macro at the beginning of a function, I'm worried that it'll be compiled away.

Reply:
I would expect it to be only used inside a conditional or an assert, so that should be ok